drivers/source/SFDP.cpp

```diff
@@ -393,34 +393,31 @@ int sfdp_find_addr_region(bd_addr_t offset, const sfdp_hdr_info &sfdp_info)
 
 }
 
-int sfdp_iterate_next_largest_erase_type(uint8_t &bitfield,
+int sfdp_iterate_next_largest_erase_type(uint8_t bitfield,
                                          int size,
                                          int offset,
                                          int region,
                                          const sfdp_smptbl_info &smptbl)
 {
     uint8_t type_mask = SFDP_ERASE_BITMASK_TYPE4;
-    int largest_erase_type = 0;
-
-    int idx;
-    for (idx = 3; idx >= 0; idx--) {
+    unsigned int erase_size;
+    for (int idx = 3; idx >= 0; idx--) {
         if (bitfield & type_mask) {
-            largest_erase_type = idx;
-            if ((size > (int)(smptbl.erase_type_size_arr[largest_erase_type])) &&
-                    ((smptbl.region_high_boundary[region] - offset)
-                     > (uint64_t)(smptbl.erase_type_size_arr[largest_erase_type]))) {
-                break;
-            } else {
-                bitfield &= ~type_mask;
+            erase_size = smptbl.erase_type_size_arr[idx];
+            // Criteria:
+            // * offset is aligned to the type's erase size
+            // * erase size is no larger than the requested size,
+            // * erase range does not exceed the region boundary
+            if ((offset % erase_size == 0) && (size >= erase_size) &&
+                    (offset + erase_size - 1 <= smptbl.region_high_boundary[region])) {
+                return idx;
             }
         }
         type_mask = type_mask >> 1;
     }
 
-    if (idx == -1) {
-        tr_error("No erase type was found for current region addr");
-    }
-    return largest_erase_type;
```

Comment:
The old implementation incorrectly returned the last erase type it checked, instead of an error.